More ptr metadata gvn

scottmcm opened the pull request:
There's basically 3 parts to this PR.

1. Allow references as arguments to `UnOp::PtrMetadata` 
 
This is a MIR semantics addition, so
r? mir

Rather than just raw pointers, also allow references to be passed to `PtrMetadata`.  That means the length of a slice can be just `PtrMetadata(_1)` instead of also needing a ref-to-pointer statement (`_2 = &raw *_1` + `PtrMetadata(_2)`).

AFAIK there should be no provenance or tagging implications of looking at the *metadata* of a pointer, and the code in the backends actually already supported it (other than a debug assert, given that they don't care about ptr vs reference, really), so we might as well allow it.

2. Simplify the argument to `PtrMetadata` in GVN

Because the specific kind of pointer-like thing isn't that important, GVN can simplify all those details away.  Things like `*const`-to-`*mut` casts and `&mut`-to-`&` reborrows are irrelevant, and skipping them lets it see more interesting things.

cc @cjgillot 

Notably, unsizing casts for arrays.  GVN supported that for `Len`, and now it sees it for `PtrMetadata` as well, allowing `PtrMetadata(pointer)` to become a constant if that pointer came from an array-to-slice unsizing, even through a bunch of other possible steps.

3. Replace `NormalizeArrayLen` with GVN

The `NormalizeArrayLen` pass hasn't been running even in optimized builds for well over a year, and it turns out that GVN -- which *is* on in optimized builds -- can do everything it was trying to do.

So the code for the pass is deleted, but the tests are kept, just changed to the different pass.

As part of this, `LowerSliceLen` was changed to emit `PtrMetadata(_1)` instead of `Len(*_1)`, a small step on the road to eventually eliminating `Rvalue::Len`.

scottmcm committed:
Replace `NormalizeArrayLen` with `GVN`

GVN is actually on in release, and covers all the same things (or more), with `LowerSliceLen` changed to produce `PtrMetadata`.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9534821800/job/26279796195) [(plain)](https://github.com/rust-lang/rust/commit/7323f2adda2dcbc3cd5b0d1ae03d2a176c73be22/checks/26279796195/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 25.5s done
#16 DONE 30.4s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---

---- [coverage-map] tests/coverage/closure_macro.rs stdout ----
diff of cov-map:

22 - Code(Counter(0)) at (prev + 3, 1) to (start + 0, 2)
24 Function name: closure_macro::main::{closure#0}
24 Function name: closure_macro::main::{closure#0}
- Raw bytes (35): 0x[01, 01, 03, 01, 05, 05, 0b, 09, 00, 05, 01, 10, 1c, 03, 21, 05, 04, 11, 01, 27, 02, 03, 11, 00, 16, 00, 00, 17, 00, 1e, 07, 02, 09, 00, 0a]
+ Raw bytes (35): 0x[01, 01, 03, 01, 05, 05, 0b, 09, 0d, 05, 01, 10, 1c, 03, 21, 05, 04, 11, 01, 27, 02, 03, 11, 00, 16, 0d, 00, 17, 00, 1e, 07, 02, 09, 00, 0a]
27 - file 0 => global file 1
28 Number of expressions: 3

29 - expression 0 operands: lhs = Counter(0), rhs = Counter(1)
29 - expression 0 operands: lhs = Counter(0), rhs = Counter(1)
30 - expression 1 operands: lhs = Counter(1), rhs = Expression(2, Add)
- - expression 2 operands: lhs = Counter(2), rhs = Zero
+ - expression 2 operands: lhs = Counter(2), rhs = Counter(3)
32 Number of file 0 mappings: 5
33 - Code(Counter(0)) at (prev + 16, 28) to (start + 3, 33)
34 - Code(Counter(1)) at (prev + 4, 17) to (start + 1, 39)

35 - Code(Expression(0, Sub)) at (prev + 3, 17) to (start + 0, 22)
36     = (c0 - c1)
- - Code(Zero) at (prev + 0, 23) to (start + 0, 30)
+ - Code(Counter(3)) at (prev + 0, 23) to (start + 0, 30)
38 - Code(Expression(1, Add)) at (prev + 2, 9) to (start + 0, 10)
-     = (c1 + (c2 + Zero))
+     = (c1 + (c2 + c3))
41 


The actual cov-map differed from the expected cov-map.
The actual cov-map differed from the expected cov-map.
Actual cov-map saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/coverage/closure_macro.coverage-map/closure_macro.cov-map
error: 1 errors occurred comparing coverage output.
status: exit status: 0
status: exit status: 0
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/coverage-dump" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/coverage/closure_macro.coverage-map/closure_macro.ll"
Function name: closure_macro::load_configuration_files
Function name: closure_macro::load_configuration_files
Raw bytes (9): 0x[01, 01, 00, 01, 01, 1d, 01, 02, 02]
- file 0 => global file 1
Number of expressions: 0
Number of file 0 mappings: 1
Number of file 0 mappings: 1
- Code(Counter(0)) at (prev + 29, 1) to (start + 2, 2)
Function name: closure_macro::main
Function name: closure_macro::main
Raw bytes (31): 0x[01, 01, 01, 01, 05, 05, 01, 21, 01, 01, 21, 02, 02, 09, 00, 0f, 05, 00, 54, 00, 55, 02, 02, 09, 02, 0b, 01, 03, 01, 00, 02]
- file 0 => global file 1
Number of expressions: 1
- expression 0 operands: lhs = Counter(0), rhs = Counter(1)
Number of file 0 mappings: 5
Number of file 0 mappings: 5
- Code(Counter(0)) at (prev + 33, 1) to (start + 1, 33)
- Code(Expression(0, Sub)) at (prev + 2, 9) to (start + 0, 15)
    = (c0 - c1)
- Code(Counter(1)) at (prev + 0, 84) to (start + 0, 85)
- Code(Expression(0, Sub)) at (prev + 2, 9) to (start + 2, 11)
    = (c0 - c1)
- Code(Counter(0)) at (prev + 3, 1) to (start + 0, 2)
Function name: closure_macro::main::{closure#0}
Function name: closure_macro::main::{closure#0}
Raw bytes (35): 0x[01, 01, 03, 01, 05, 05, 0b, 09, 0d, 05, 01, 10, 1c, 03, 21, 05, 04, 11, 01, 27, 02, 03, 11, 00, 16, 0d, 00, 17, 00, 1e, 07, 02, 09, 00, 0a]
- file 0 => global file 1
Number of expressions: 3
- expression 0 operands: lhs = Counter(0), rhs = Counter(1)
- expression 0 operands: lhs = Counter(0), rhs = Counter(1)
- expression 1 operands: lhs = Counter(1), rhs = Expression(2, Add)
- expression 2 operands: lhs = Counter(2), rhs = Counter(3)
Number of file 0 mappings: 5
- Code(Counter(0)) at (prev + 16, 28) to (start + 3, 33)
- Code(Counter(1)) at (prev + 4, 17) to (start + 1, 39)
- Code(Expression(0, Sub)) at (prev + 3, 17) to (start + 0, 22)
    = (c0 - c1)
- Code(Counter(3)) at (prev + 0, 23) to (start + 0, 30)
- Code(Expression(1, Add)) at (prev + 2, 9) to (start + 0, 10)
    = (c1 + (c2 + c3))
stderr: none


---- [coverage-map] tests/coverage/closure_macro_async.rs stdout ----
---- [coverage-map] tests/coverage/closure_macro_async.rs stdout ----
diff of cov-map:

30 - Code(Counter(0)) at (prev + 3, 1) to (start + 0, 2)
32 Function name: closure_macro_async::test::{closure#0}::{closure#0}
32 Function name: closure_macro_async::test::{closure#0}::{closure#0}
- Raw bytes (35): 0x[01, 01, 03, 01, 05, 05, 0b, 09, 00, 05, 01, 12, 1c, 03, 21, 05, 04, 11, 01, 27, 02, 03, 11, 00, 16, 00, 00, 17, 00, 1e, 07, 02, 09, 00, 0a]
+ Raw bytes (35): 0x[01, 01, 03, 01, 05, 05, 0b, 09, 0d, 05, 01, 12, 1c, 03, 21, 05, 04, 11, 01, 27, 02, 03, 11, 00, 16, 0d, 00, 17, 00, 1e, 07, 02, 09, 00, 0a]
35 - file 0 => global file 1
36 Number of expressions: 3

37 - expression 0 operands: lhs = Counter(0), rhs = Counter(1)
37 - expression 0 operands: lhs = Counter(0), rhs = Counter(1)
38 - expression 1 operands: lhs = Counter(1), rhs = Expression(2, Add)
- - expression 2 operands: lhs = Counter(2), rhs = Zero
+ - expression 2 operands: lhs = Counter(2), rhs = Counter(3)
40 Number of file 0 mappings: 5
41 - Code(Counter(0)) at (prev + 18, 28) to (start + 3, 33)
42 - Code(Counter(1)) at (prev + 4, 17) to (start + 1, 39)

43 - Code(Expression(0, Sub)) at (prev + 3, 17) to (start + 0, 22)
44     = (c0 - c1)
- - Code(Zero) at (prev + 0, 23) to (start + 0, 30)
+ - Code(Counter(3)) at (prev + 0, 23) to (start + 0, 30)
46 - Code(Expression(1, Add)) at (prev + 2, 9) to (start + 0, 10)
-     = (c1 + (c2 + Zero))
+     = (c1 + (c2 + c3))
49 


The actual cov-map differed from the expected cov-map.
The actual cov-map differed from the expected cov-map.
Actual cov-map saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/coverage/closure_macro_async.coverage-map/closure_macro_async.cov-map
error: 1 errors occurred comparing coverage output.
status: exit status: 0
status: exit status: 0
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/coverage-dump" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/coverage/closure_macro_async.coverage-map/closure_macro_async.ll"
Function name: closure_macro_async::load_configuration_files
Function name: closure_macro_async::load_configuration_files
Raw bytes (9): 0x[01, 01, 00, 01, 01, 1f, 01, 02, 02]
- file 0 => global file 1
Number of expressions: 0
Number of file 0 mappings: 1
Number of file 0 mappings: 1
- Code(Counter(0)) at (prev + 31, 1) to (start + 2, 2)
Function name: closure_macro_async::test
Function name: closure_macro_async::test
Raw bytes (9): 0x[01, 01, 00, 01, 01, 23, 01, 00, 2b]
- file 0 => global file 1
Number of expressions: 0
Number of file 0 mappings: 1
Number of file 0 mappings: 1
- Code(Counter(0)) at (prev + 35, 1) to (start + 0, 43)
Function name: closure_macro_async::test::{closure#0}
Function name: closure_macro_async::test::{closure#0}
Raw bytes (31): 0x[01, 01, 01, 01, 05, 05, 01, 23, 2b, 01, 21, 02, 02, 09, 00, 0f, 05, 00, 54, 00, 55, 02, 02, 09, 02, 0b, 01, 03, 01, 00, 02]
- file 0 => global file 1
Number of expressions: 1
- expression 0 operands: lhs = Counter(0), rhs = Counter(1)
Number of file 0 mappings: 5
Number of file 0 mappings: 5
- Code(Counter(0)) at (prev + 35, 43) to (start + 1, 33)
- Code(Expression(0, Sub)) at (prev + 2, 9) to (start + 0, 15)
    = (c0 - c1)
- Code(Counter(1)) at (prev + 0, 84) to (start + 0, 85)
- Code(Expression(0, Sub)) at (prev + 2, 9) to (start + 2, 11)
    = (c0 - c1)
- Code(Counter(0)) at (prev + 3, 1) to (start + 0, 2)
Function name: closure_macro_async::test::{closure#0}::{closure#0}
Function name: closure_macro_async::test::{closure#0}::{closure#0}
Raw bytes (35): 0x[01, 01, 03, 01, 05, 05, 0b, 09, 0d, 05, 01, 12, 1c, 03, 21, 05, 04, 11, 01, 27, 02, 03, 11, 00, 16, 0d, 00, 17, 00, 1e, 07, 02, 09, 00, 0a]
- file 0 => global file 1
Number of expressions: 3
- expression 0 operands: lhs = Counter(0), rhs = Counter(1)
- expression 0 operands: lhs = Counter(0), rhs = Counter(1)
- expression 1 operands: lhs = Counter(1), rhs = Expression(2, Add)
- expression 2 operands: lhs = Counter(2), rhs = Counter(3)
Number of file 0 mappings: 5
- Code(Counter(0)) at (prev + 18, 28) to (start + 3, 33)
- Code(Counter(1)) at (prev + 4, 17) to (start + 1, 39)
- Code(Expression(0, Sub)) at (prev + 3, 17) to (start + 0, 22)
    = (c0 - c1)
- Code(Counter(3)) at (prev + 0, 23) to (start + 0, 30)
- Code(Expression(1, Add)) at (prev + 2, 9) to (start + 0, 10)
    = (c1 + (c2 + c3))
stderr: none



```

</details>

scottmcm committed:
I'd never even heard of a coverage map

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 72cda61f63f9082949b6dd14daea4db58e33cb96 with merge bcb40db9974a93769468e3683e983c8a2da624cf...
<!-- homu: {"type":"TryBuildStarted","head_sha":"72cda61f63f9082949b6dd14daea4db58e33cb96","merge_sha":"bcb40db9974a93769468e3683e983c8a2da624cf"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9535181158/job/26281802355)
Build commit: bcb40db9974a93769468e3683e983c8a2da624cf (`bcb40db9974a93769468e3683e983c8a2da624cf`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9535181158/job/26281802355"},"merge_sha":"bcb40db9974a93769468e3683e983c8a2da624cf"} -->

rust-timer commented:
Queued bcb40db9974a93769468e3683e983c8a2da624cf with parent 12b33d36f3de0fb30ca08c49f5b2f4fa111861a4, future [comparison URL](https://perf.rust-lang.org/compare.html?start=12b33d36f3de0fb30ca08c49f5b2f4fa111861a4&end=bcb40db9974a93769468e3683e983c8a2da624cf).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.2 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (bcb40db9974a93769468e3683e983c8a2da624cf): [comparison URL](https://perf.rust-lang.org/compare.html?start=12b33d36f3de0fb30ca08c49f5b2f4fa111861a4&end=bcb40db9974a93769468e3683e983c8a2da624cf&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=12b33d36f3de0fb30ca08c49f5b2f4fa111861a4&end=bcb40db9974a93769468e3683e983c8a2da624cf&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.7%  | [0.7%, 0.7%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 0.3%  | [0.2%, 0.5%]   | 5     |
| Improvements ✅ <br /> (primary)   | -0.5% | [-1.1%, -0.3%] | 8     |
| Improvements ✅ <br /> (secondary) | -2.0% | [-2.0%, -2.0%] | 1     |
| All ❌✅ (primary)                 | -0.4% | [-1.1%, 0.7%]  | 9     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=12b33d36f3de0fb30ca08c49f5b2f4fa111861a4&end=bcb40db9974a93769468e3683e983c8a2da624cf&stat=max-rss)
<details>
<summary>Results (primary -1.2%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range           | count |
|:----------------------------------:|:-----:|:---------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 5.2%  | [4.5%, 5.9%]    | 2     |
| Regressions ❌ <br /> (secondary)  | -     | -               | 0     |
| Improvements ✅ <br /> (primary)   | -5.5% | [-10.2%, -2.9%] | 3     |
| Improvements ✅ <br /> (secondary) | -     | -               | 0     |
| All ❌✅ (primary)                 | -1.2% | [-10.2%, 5.9%]  | 5     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=12b33d36f3de0fb30ca08c49f5b2f4fa111861a4&end=bcb40db9974a93769468e3683e983c8a2da624cf&stat=cycles:u)
<details>
<summary>Results (primary 0.3%, secondary 7.0%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.3%  | [1.3%, 1.3%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 7.0%  | [3.9%, 8.9%]   | 4     |
| Improvements ✅ <br /> (primary)   | -0.7% | [-0.7%, -0.7%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.3%  | [-0.7%, 1.3%]  | 2     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=12b33d36f3de0fb30ca08c49f5b2f4fa111861a4&end=bcb40db9974a93769468e3683e983c8a2da624cf&stat=size:linked_artifact)
<details>
<summary>Results (primary 0.0%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.1%  | [0.0%, 0.7%]   | 13    |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.1% | [-0.4%, -0.0%] | 19    |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.0%  | [-0.4%, 0.7%]  | 32    |
</details>

**Bootstrap**: 672.776s -> 674.201s (0.21%)
**Artifact size**: 319.83 MiB -> 320.43 MiB (0.19%)

cjgillot commented on compiler/rustc_mir_transform/src/gvn.rs:
```diff
@@ -840,8 +840,56 @@ impl<'body, 'tcx> VnState<'body, 'tcx> {
+                        }
+                        break;
+                    }
+                }
```
Can all this go in `simplify_unary`?

cjgillot commented on compiler/rustc_mir_transform/src/gvn.rs:
```diff
@@ -840,8 +840,56 @@ impl<'body, 'tcx> VnState<'body, 'tcx> {
+                            from,
+                            to,
+                        } = value
+                            && from.builtin_deref(true) == to.builtin_deref(true)
```
Could you add a comment for this check?
Should we handle the case of both being slices of different element types?

cjgillot commented on compiler/rustc_mir_transform/src/gvn.rs:
```diff
@@ -840,8 +840,56 @@ impl<'body, 'tcx> VnState<'body, 'tcx> {
+                        let value = self.get(arg);
+
+                        // FIXME: remove MutToConst after #126308
+                        if let Value::Cast {
```
Could this be a match instead of several `if let`?

RalfJung commented on compiler/rustc_mir_transform/src/validate.rs:
```diff
@@ -1116,12 +1116,17 @@ impl<'a, 'tcx> Visitor<'tcx> for TypeChecker<'a, 'tcx> {
                             self.fail(location, "PtrMetadata should be in runtime MIR only");
                         }
 
-                        check_kinds!(a, "Cannot PtrMetadata non-pointer type {:?}", ty::RawPtr(..));
```
The [doc comment](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/enum.UnOp.html#variant.PtrMetadata) should also be updated.

scottmcm committed:
Add a try_as_constant+try_as_local helper

No behaviour changes.

scottmcm committed:
[GVN] Add tests for generic pointees with PtrMetadata

bors commented:
:umbrella: The latest upstream changes (presumably #126308) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

scottmcm committed:
More GVN for PtrMetadata

`PtrMetadata` doesn't care about `*const`/`*mut`/`&`/`&mut`, so GVN away those casts in its argument.

This includes updating MIR to allow calling PtrMetadata on references too, not just raw pointers.  That means that `[T]::len` can be just `_0 = PtrMetadata(_1)`, for example.

# Conflicts:
#	tests/mir-opt/pre-codegen/slice_index.slice_get_unchecked_mut_range.PreCodegen.after.panic-abort.mir
#	tests/mir-opt/pre-codegen/slice_index.slice_get_unchecked_mut_range.PreCodegen.after.panic-unwind.mir

cjgillot commented on tests/mir-opt/gvn.rs:
```diff
@@ -816,6 +818,62 @@ fn casts_before_aggregate_raw_ptr(x: *const u32) -> *const [u8] {
+    std::intrinsics::ptr_metadata(x)
+}
+
+#[custom_mir(dialect = "runtime")]
```
This test is a bit tricky, could you add a small explanation why each case must work as shown?

cjgillot commented on compiler/rustc_mir_transform/src/gvn.rs:
```diff
@@ -979,8 +967,62 @@ impl<'body, 'tcx> VnState<'body, 'tcx> {
+                    // that's not known, then matching tail types is also ok,
+                    // allowing things like `*mut (?A, ?T)` <-> `*mut (?B, ?T)`.
+                    Value::Cast { kind: CastKind::PtrToPtr, value: inner, from, to }
+                        if from.builtin_deref(true)?.ptr_metadata_ty_or_tail(self.tcx, |t| t)
+                            == to.builtin_deref(true)?.ptr_metadata_ty_or_tail(self.tcx, |t| t) =>
```
Should we pass a normalization routine with our param_env? Or is it just being pedantic?

cjgillot commented on compiler/rustc_mir_transform/src/gvn.rs:
```diff
@@ -992,9 +1034,26 @@ impl<'body, 'tcx> VnState<'body, 'tcx> {
+                    ..
+                },
+            ) if let ty::Slice(..) = to.builtin_deref(true).unwrap().kind()
+                && let ty::Array(_, len) = from.builtin_deref(true).unwrap().kind() =>
```
Nit: why do we `unwrap()` the result of builtin_deref here, while we `?` earlier?

cjgillot reviewed (commented):
Some nits on the mir-opt changes. r=me on those when the mir semantics addition is approved.

RalfJung commented:
> Allow references as arguments to UnOp::PtrMetadata

This seems entirely uncontroversial to me from a MIR semantics / opsem perspective, so fine for me.

bors commented:
:pushpin: Commit 55d13379ac57227a08ba537316adcf2d3c388f57 has been approved by `cjgillot`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cjgillot 55d13379ac57227a08ba537316adcf2d3c388f57 -->
<!-- homu: {"type":"Approved","sha":"55d13379ac57227a08ba537316adcf2d3c388f57","approver":"cjgillot","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 55d13379ac57227a08ba537316adcf2d3c388f57 with merge e32ea4822ba3f7c9054a7fb5345232388cc50c4b...
<!-- homu: {"type":"BuildStarted","head_sha":"55d13379ac57227a08ba537316adcf2d3c388f57","merge_sha":"e32ea4822ba3f7c9054a7fb5345232388cc50c4b"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9609769629/job/26509673350)
Approved by: cjgillot
Pushing e32ea4822ba3f7c9054a7fb5345232388cc50c4b to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"cjgillot","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9609769629/job/26509673350"},"merge_sha":"e32ea4822ba3f7c9054a7fb5345232388cc50c4b"} -->

rust-timer commented:
Finished benchmarking commit (e32ea4822ba3f7c9054a7fb5345232388cc50c4b): [comparison URL](https://perf.rust-lang.org/compare.html?start=4e6de37349c7838df095b085a2640cb6a007c96a&end=e32ea4822ba3f7c9054a7fb5345232388cc50c4b&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=4e6de37349c7838df095b085a2640cb6a007c96a&end=e32ea4822ba3f7c9054a7fb5345232388cc50c4b&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.2%  | [0.2%, 0.2%]   | 1     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.0% | [-1.4%, -0.6%] | 2     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.6% | [-1.4%, 0.2%]  | 3     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=4e6de37349c7838df095b085a2640cb6a007c96a&end=e32ea4822ba3f7c9054a7fb5345232388cc50c4b&stat=max-rss)
<details>
<summary>Results (primary 0.8%, secondary 3.9%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 7.7%  | [7.7%, 7.7%]   | 2     |
| Regressions ❌ <br /> (secondary)  | 3.9%  | [2.0%, 5.6%]   | 4     |
| Improvements ✅ <br /> (primary)   | -3.8% | [-5.5%, -0.5%] | 3     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.8%  | [-5.5%, 7.7%]  | 5     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=4e6de37349c7838df095b085a2640cb6a007c96a&end=e32ea4822ba3f7c9054a7fb5345232388cc50c4b&stat=cycles:u)
<details>
<summary>Results (primary -1.4%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.4% | [-1.5%, -1.4%] | 2     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -1.4% | [-1.5%, -1.4%] | 2     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=4e6de37349c7838df095b085a2640cb6a007c96a&end=e32ea4822ba3f7c9054a7fb5345232388cc50c4b&stat=size:linked_artifact)
<details>
<summary>Results (primary 0.0%, secondary 0.3%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.1%  | [0.0%, 0.4%]   | 13    |
| Regressions ❌ <br /> (secondary)  | 0.3%  | [0.3%, 0.3%]   | 3     |
| Improvements ✅ <br /> (primary)   | -0.1% | [-0.2%, -0.0%] | 15    |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.0%  | [-0.2%, 0.4%]  | 28    |
</details>

**Bootstrap**: 698.008s -> 696.377s (-0.23%)
**Artifact size**: 326.58 MiB -> 326.82 MiB (0.07%)

nnethercote commented:
Some small wins and losses in each category, which roughly balance out.

@rustbot label: +perf-regression-triaged
